### PR TITLE
feat(rest): provide more options to customize api explorer and openapi spec

### PR DESCRIPTION
**Pull request by raymondfeng**
The generated openapi spec has a dummy server url at the moment: `servers: [url: '/']`. The PR adds more options to rest server config.

See https://github.com/strongloop/loopback-next/blob/c78d636b9cccf66bc843ec49b203b0a6c319f78e/packages/rest/README.md#configuration

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

**Comment by bajtos** [comment]
What's wrong with using a path only in the server url (`servers: [url: '/']`)? This is not a dummy value but an intentional design that prevents all kinds of issues in deployments where the application does not know its public hostname and port.

Please note that OpenAPI spec allows relative URLs like the one we are using. See https://github.com/OAI/OpenAPI-Specification/blob/master/versions/3.0.1.md#serverObject

> `url` - `string` - `REQUIRED`
> A URL to the target host. This URL supports Server Variables and MAY be relative, to indicate that the host location is relative to the location where the OpenAPI document is being served. Variable substitutions will be made when a variable is named in {brackets}.

I am personally against landing this change and prefer to look into ways how to fix OpenAPI consumers to correctly support relative URLs.

**Comment by raymondfeng** [comment]
@bajtos Good point. Let's evaluate a bit more.

**Comment by virkt25** [comment]
From what I remember the OASGraph team said this was not a critical fix unlike the `responses` field and `required: true` for path parameters.

**Comment by bajtos** [comment]
Can we perhaps make the server URL configurable?

 - By default, use a relative URL as we do now (`/`)
 - Add an option to use the URL inferred from the app config, the request properties and `X-Forwarded-*` headers. 
 - Add an option (perhaps the same one as above?) allowing the user to provide their own protocol+hostname+port.

Thoughts?

**Comment by raymondfeng** [comment]
> By default, use a relative URL as we do now (/)
+1.
> Add an option to use the URL inferred from the app config, the request properties and X-Forwarded-* headers.
+1.
> Add an option (perhaps the same one as above?) allowing the user to provide their own protocol+hostname+port.
`server.api()` allows the customization of the initial openapi spec, which can include `servers`.

**Comment by raymondfeng** [comment]
@bajtos @virkt25 I have introduced more options for the REST server. PTAL.

**Comment by raymondfeng** [comment]
@bajtos PTAL

File: packages/rest/README.md [hunk 1]
```diff
@@ -52,16 +52,35 @@ Application options.
...
+    // Optional configuration for API explorer and spec
+    apiExplorer: {
+      // URL for the hosted API Explorer UI
+      // default to https://loopback.io/api-explorer
+      explorerUrl: 'http://petstore.swagger.io',
+      // Set `servers` based on HTTP request headers, default to `false`
+      setServersFromRequest: false,
+      // Template the OpenAPI spec, which will be filled with `paths` from
+      // controllers.
+      openApiSpec: {
```
Comment (virkt25) [inline packages/rest/README.md]: I think the base template for OpenAPI spec should not be set in the `apiExplorer` object. It should be it's own top level property under `rest`.

Comment (raymondfeng) [reply packages/rest/README.md]: Good point. Fixed.

File: packages/rest/README.md [hunk 2]
```diff
@@ -52,16 +52,36 @@ Application options.
...
-| -------- | --------------- | --------------------------------------------------------------------------------------------------------- |
-| port     | number          | Specify the port on which the RestServer will listen for traffic.                                         |
-| sequence | SequenceHandler | Use a custom SequenceHandler to change the behavior of the RestServer for the request-response lifecycle. |
+| Property    | Type               | Purpose                                                                                                   |
+| ----------- | ------------------ | --------------------------------------------------------------------------------------------------------- |
+| port        | number             | Specify the port on which the RestServer will listen for traffic.                                         |
+| sequence    | SequenceHandler    | Use a custom SequenceHandler to change the behavior of the RestServer for the request-response lifecycle. |
+| apiExplorer | ApiExplorerOptions | Custom how API explorer and OpenAPI spec is served                                                        |
```
Comment (virkt25) [inline packages/rest/README.md]: Please document `openApiSpec` property in this table. :)

File: packages/rest/README.md [hunk 3]
```diff
@@ -52,16 +52,36 @@ Application options.
...
+      servers: [{url: 'http://127.0.0.1:8080'}],
+    },
+
+    // Optional configuration for API explorer and spec
+    apiExplorer: {
+      // URL for the hosted API Explorer UI
+      // default to https://loopback.io/api-explorer
+      explorerUrl: 'http://petstore.swagger.io',
+      // Set `servers` based on HTTP request headers, default to `false`
+      setServersFromRequest: false,
```
Comment (virkt25) [inline packages/rest/README.md]: I don't think this is a `apiExplorer` related property either but now sure what a better place for this might be.

Comment (bajtos) [reply packages/rest/README.md]: How about moving this property from `rest.apiExplorer.setServersFromRequest` to something like `rest.updateOpenApiServersFromRequest`?

File: packages/rest/README.md [hunk 4]
```diff
@@ -52,16 +52,37 @@ Application options.
 const app = new RestApplication({
   rest: {
     port: 3001,
+
+    // Template of the OpenAPI spec, which will be filled with `paths` from
+    // controllers.
+    openApiSpec: {
```
Comment (bajtos) [inline packages/rest/README.md]: I believe we already have an API for setting the initial OpenAPI spec values, see http://apidocs.loopback.io/@loopback%2fdocs/rest.html#RestApplication.prototype.api

Can we keep using `app.api()` instead of introducing this new option? If not, then what is your reasoning? What benefits do you see in this declarative approach?

The current downside of `app.api()` is that the user has to provide `spec.openapi` and `spec.paths` properties in addition to `info`.

```ts
app.api({
  openapi: '3.0.1',
  info: { /*...*/ },
  servers: [/*...*/],
  paths: {},
});
```

On the brighter side, `app.api()` allows TypeScript to verify the correctness of the passed values. I am not sure if the type checks work for your declarative approach, because the type used for options allow arbitrary (additional) properties.

File: packages/rest/README.md [hunk 5]
```diff
@@ -52,16 +52,37 @@ Application options.
...
+        version: '1.0.0',
+      },
+      servers: [{url: 'http://127.0.0.1:8080'}],
+    },
+
+    // Optional configuration for API explorer
+    apiExplorer: {
+      // URL for the hosted API Explorer UI
+      // default to https://loopback.io/api-explorer
+      explorerUrl: 'http://petstore.swagger.io',
```
Comment (bajtos) [inline packages/rest/README.md]: Can we use a shorter name `url`? The fact that this setting is controlling API Explorer is already encoded in the parent property name `apiExplorer`.

File: packages/rest/README.md [hunk 6]
```diff
@@ -52,16 +52,37 @@ Application options.
...
-| -------- | --------------- | --------------------------------------------------------------------------------------------------------- |
-| port     | number          | Specify the port on which the RestServer will listen for traffic.                                         |
-| sequence | SequenceHandler | Use a custom SequenceHandler to change the behavior of the RestServer for the request-response lifecycle. |
+| Property    | Type               | Purpose                                                                                                   |
+| ----------- | ------------------ | --------------------------------------------------------------------------------------------------------- |
+| port        | number             | Specify the port on which the RestServer will listen for traffic.                                         |
+| sequence    | SequenceHandler    | Use a custom SequenceHandler to change the behavior of the RestServer for the request-response lifecycle. |
+| openApiSpec | OpenApiSpec        | Base template of the OpenAPI spec                                                                         |
```
Comment (bajtos) [inline packages/rest/README.md]: This is not entirely correct. `OpenApiSpec` type requires `openapi` version and `paths` object to be present.

File: packages/rest/src/rest.server.ts [hunk 7]
```diff
@@ -123,6 +123,7 @@ export class RestServer extends Context implements Server, HttpServerLike {
    */
   public requestHandler: HttpRequestListener;
 
+  protected _config: RestServerConfig;
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: IMO, this property should not be modified after the constructor has finished, let's mark it as `readonly` please.

Is there any particular reason for keeping the config protected? I mean can we make it `public`?

File: packages/rest/src/rest.server.ts [hunk 8]
```diff
@@ -336,11 +334,19 @@ export class RestServer extends Context implements Server, HttpServerLike {
   private async _serveOpenApiSpec(
     request: Request,
     response: Response,
-    options?: OpenApiSpecOptions,
+    settings?: OpenApiSpecOptions,
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: This looks suspicious to me - a `settings` argument having type `OpenApiSpecOptions`? Let's revert this change or use a more descriptive name, e.g. `specOptions`.

File: packages/rest/src/rest.server.ts [hunk 9]
```diff
@@ -336,11 +334,19 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
   ) {
-    options = options || {version: '3.0.0', format: 'json'};
+    settings = settings || {version: '3.0.0', format: 'json'};
     let specObj = this.getApiSpec();
-    if (options.format === 'json') {
+    if (
+      this._config.apiExplorer &&
+      this._config.apiExplorer.setServersFromRequest
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: To avoid repetitive check for `this._config.apiExplorer`, I am proposing to always initialize this property in the constructor, so that it's either an empty object or the user-provider object. Then we can simplify this part as follows:

```ts
if (this._config.apiExplorer.setServersFromRequest)
```

See also https://loopback.io/doc/en/contrib/style-guide.html#indentation-of-multi-line-expressions-in-if

File: packages/rest/src/rest.server.ts [hunk 10]
```diff
@@ -380,17 +386,11 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
-    options: RestServerConfig,
-  ) {
+  private async _redirectToSwaggerUI(request: Request, response: Response) {
     const baseUrl =
-      options.apiExplorerUrl || 'https://loopback.io/api-explorer';
-    const openApiUrl = `${this._getUrlForClient(
-      request,
-      options,
-    )}/openapi.json`;
+      (this._config.apiExplorer && this._config.apiExplorer.explorerUrl) ||
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: To avoid repetitive check for `this._config.apiExplorer`, I am proposing to always initialize this property in the constructor, so that it's either an empty object or the user-provider object. Then we can simplify this part as follows:

```ts
const baseUrl = this._config.apiExplorer.explorerUrl || 'https://loopback.io/api-explorer';
```

We can even set `explorerUrl` to the default value in the constructor, that would allow us to simplify this code even further.

File: packages/rest/test/integration/rest.server.integration.ts [hunk 11]
```diff
@@ -120,7 +126,11 @@ describe('RestServer (integration)', () => {
   });
 
   it('exposes "GET /openapi.json" endpoint', async () => {
-    const server = await givenAServer({rest: {port: 0}});
+    const server = await givenAServer({
+      rest: {
+        port: 0,
+      },
+    });
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: This looks like a change of formatting only, can we revert it please? Unless there is a reason for this change, in which case I'd like to hear it :)

File: packages/rest/test/integration/rest.server.integration.ts [hunk 12]
```diff
@@ -159,6 +172,110 @@ describe('RestServer (integration)', () => {
...
+                },
+                description: 'greeting of the day',
+              },
+            },
+          },
+        },
+      },
+    });
+    expect(response.get('Access-Control-Allow-Origin')).to.equal('*');
+    expect(response.get('Access-Control-Allow-Credentials')).to.equal('true');
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: IIUC, this test is verifying how `rest.openApiSpec` config is applied. There is no need to check CORS headers in such case, they have been already verified by previous tests.

Please remove these two assertions.

File: packages/rest/test/integration/rest.server.integration.ts [hunk 13]
```diff
@@ -159,6 +172,110 @@ describe('RestServer (integration)', () => {
...
+                description: 'greeting of the day',
+              },
+            },
+          },
+        },
+      },
+    });
+    expect(response.body.servers[0].url).to.match(/http:\/\/127.0.0.1\:\d+/);
+    expect(response.get('Access-Control-Allow-Origin')).to.equal('*');
+    expect(response.get('Access-Control-Allow-Credentials')).to.equal('true');
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: Ditto. 

IIUC, this test is verifying how `rest.openApiSpec` config is applied. There is no need to check CORS headers in such case, they have been already verified by previous tests.

Please remove these two assertions.

File: packages/rest/README.md (hunk 4 above)
Comment (bajtos) [reply packages/rest/README.md]: We can relax the constraints on the argument of `app.api()` to allow any subset of OpenAPISpec and fill in `openapi` version and an empty `paths` object ourselves.

**Comment by bajtos** [comment]
On the second thought, my main objection against providing OpenAPI spec in `RestServerConfig` (and `ApplicationConfig`) (see https://github.com/strongloop/loopback-next/pull/1637#discussion_r213998478) is that the configuration of application behavior/operations should not be something that server/application users are allowed to change!

This something we have already discussed in the past, see https://github.com/strongloop/loopback-next/issues/742.

> Right now, Application constructor accepts a config/options object that mixes two very different concerns:
>
> - The functionality/features the application should implement - what controllers it contains, what repositories are registered, etc. ("the assembly instructions")
> - The configuration - what port to listen on, what database servers to connect to, etc.
> 
> IMO, the consumers of an application must not change its functionality and features (with the exception of enabling/disabling explicit feature flags) and therefore the config/options argument should contain only the configuration (port numbers, connection strings, feature flags).
> 
> The ApplicationConfig class should contain only things like HTTP port and database connection strings, there should be no entries related to what component, controller, repositories (etc.) the application is using. User application classes should call this.controller(), this.repository() (etc.) to register their artifacts, and eventually use the new bootstrapper module.

In this light, if we want to make OpenAPI `servers` field configurable via app/server options, then only the array of `ServerObject`s should be accepted. For example as `rest.servers` (instead of `rest.openApi.servers`.

File: packages/rest/README.md (hunk 4 above)
Comment (bajtos) [reply packages/rest/README.md]: See also https://github.com/strongloop/loopback-next/pull/1637#issuecomment-417295302

File: packages/rest/README.md (hunk 3 above)
Comment (bajtos) [reply packages/rest/README.md]: Uh, `rest.updateOpenApiServersFromRequest` is awfully long. How about `rest.setServerUrlsFromRequest`?

File: packages/rest/README.md (hunk 4 above)
Comment (virkt25) [reply packages/rest/README.md]: Doesn't `app.api()` take in an entire spec vs. the new `openApiSpec` property which is intended to accept a base template for the spec rather than the complete spec itself (which would come from `app.api`)?

File: packages/rest/README.md (hunk 3 above)
Comment (hacksparrow) [reply packages/rest/README.md]: Or how about just `rest.serverUrlsFromRequest`?

File: packages/rest/README.md (hunk 4 above)
Comment (bajtos) [reply packages/rest/README.md]: I believe our current code merges the following sources:
 - the spec provided by `app.api()`  (if any)
 - specs provided at controller level via `@api` decorator,
 - specs built from controller-method-level decorators like `@get` and `@param`
 - specs provided via `app.route(new Route(verb, path, spec, ...))`

Comment (bajtos) [reply packages/rest/README.md]: I don't mind providing a declarative way for providing a base template for the OpenAPI spec, perhaps in addition to `app.api()` API. However, I think we need to move that declarative config to a different place than application/server config, see https://github.com/strongloop/loopback-next/pull/1637#issuecomment-417295302.

File: packages/rest/src/rest.server.ts (hunk 9 above)
Comment (raymondfeng) [reply packages/rest/src/rest.server.ts]: The style/formatting is enforced by `prettier`.

File: packages/rest/test/integration/rest.server.integration.ts (hunk 11 above)
Comment (raymondfeng) [reply packages/rest/test/integration/rest.server.integration.ts]: I use `prettier` to format the code

**Comment by raymondfeng** [comment]
@bajtos I made some changes based on your feedback. The `config` still allows `rest.openApiSpec.template` but I constrain it from taking `paths` and `components` so that it becomes a pure configuration object. PTAL.

File: packages/rest/test/integration/rest.server.integration.ts (hunk 11 above)
Comment (bajtos) [reply packages/rest/test/integration/rest.server.integration.ts]: I believe prettier allow multiple formatting styles these days, depending on how you write the code. I.e. it preserves one-liner `{rest: {port: 0}}` if you started with it, or keeps multi-line version if that's what the developer wrote.

File: packages/rest/src/rest.server.ts [hunk 14]
```diff
@@ -380,17 +370,11 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
-  ) {
+  private async _redirectToSwaggerUI(request: Request, response: Response) {
     const baseUrl =
-      options.apiExplorerUrl || 'https://loopback.io/api-explorer';
-    const openApiUrl = `${this._getUrlForClient(
-      request,
-      options,
-    )}/openapi.json`;
+      (this.config.apiExplorer && this.config.apiExplorer.url) ||
+      'https://loopback.io/api-explorer';
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: Can you set the default value for `apiExplorer.url` inside `RestServer` constructor please? Not only it will simplify the code here, but it will make it easier for 3rd parties to introspect the API Explorer configuration and obtain the actual URL used even when no custom value was configured by the user.

File: packages/rest/README.md [hunk 15]
```diff
@@ -52,16 +52,44 @@ Application options.
 const app = new RestApplication({
   rest: {
     port: 3001,
+
+    // Template of the OpenAPI spec, which will be filled with `paths` from
```
Comment (bajtos) [inline packages/rest/README.md]: I am proposing to move this new documentation to https://github.com/strongloop/loopback-next/blob/master/docs/site/Server.md and modify this README with a link pointing to that doc page.

For example, configuration of HTTPS is already described in `Server.md` only, see [here](https://github.com/strongloop/loopback-next/blob/ac011f80b229b5dd5c69851ab14b39705b852629/docs/site/Server.md#enable-https).

File: packages/rest/src/rest.server.ts [hunk 16]
```diff
@@ -703,9 +714,66 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
+   * The url for hosted API explorer UI
+   * default to https://loopback.io/api-explorer
+   */
+  url?: string;
+  /**
+   * URL for the API explorer served over plain http to deal with mixed content
+   * security imposed by browsers as the spec is exposed over `http` by default.
+   * https://github.com/strongloop/loopback-next/issues/1603
+   */
+  urlForHttp?: string;
```
Comment (bajtos) [inline packages/rest/src/rest.server.ts]: I would expect LoopBack to be smart enough to figure out that `urlForHttp` is the same as `url`, but using `http://` instead of `https://`.

What are the use cases you have in mind where `urlForHttp` would significantly differ from `url`?

File: packages/rest/README.md [hunk 17]
```diff
@@ -52,16 +52,44 @@ Application options.
 const app = new RestApplication({
   rest: {
     port: 3001,
+
+    // Optional configuration for how to serve OpenAPI spec
+    openApiSpec: {
```
Comment (bajtos) [inline packages/rest/README.md]: IMO, this configuration example has too many comments to make it practical. That's why I proposed to move this content to https://github.com/strongloop/loopback-next/blob/master/docs/site/Server.md, where you can split the code into multiple sections/paragraph and use regular text instead of code comments.

I am fine with your proposal to leave consolidation of README out of scope of this pull request.

What I would like to ask you is to move this _new README content_ to `Server.md` right now, so that we don't introduce more work for the future consolidation effort; plus reformat the documentation of config options into a form that's easier to read. It would be great to give a wider context to the readers while you are at this, e.g. why and when to change `openApiSpec` configuration, etc.

Comment (raymondfeng) [reply packages/rest/README.md]: I'll clean up the docs.

File: packages/rest/src/rest.server.ts (hunk 16 above)
Comment (raymondfeng) [reply packages/rest/src/rest.server.ts]: `urlForHttp` is optional and we can infer it from `url` in most cases. I added it there since we logically need two values and have to store both of them anyway.

File: packages/rest/test/integration/rest.server.integration.ts [hunk 18]
```diff
@@ -359,12 +480,46 @@ servers:
...
+      [
+        'https://petstore.swagger.io',
+        '\\?url=http://\\d+.\\d+.\\d+.\\d+:\\d+/openapi.json',
+      ].join(''),
+    );
+    expect(response.get('Location')).match(expectedUrl);
+  });
+
+  it('exposes "GET /swagger-ui" endpoint with apiExplorer.urlForHttp', async () => {
+    const server = await givenAServer({
```
[the remark this answers is not in the document]
Comment (raymondfeng) [reply packages/rest/test/integration/rest.server.integration.ts]: See below:
```
const expectedUrl = new RegExp(
      [
        'http://petstore.swagger.io',
        '\\?url=http://\\d+.\\d+.\\d+.\\d+:\\d+/openapi.json',
      ].join(''),
    );
    expect(response.get('Location')).match(expectedUrl);
```

The url is now `'http://petstore.swagger.io`

**Comment by raymondfeng** [comment]
@bajtos Moved REST server configuration to `docs/site/Server.md`. PTAL.

File: packages/rest/src/rest.server.ts [hunk 19]
```diff
@@ -227,32 +231,25 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
       // NOTE(bajtos) Regular routes are handled through Sequence.
       // IMO, this built-in endpoint should not run through a Sequence,
       // because it's not part of the application API itself.
       // E.g. if the app implements access/audit logs, I don't want
       // this endpoint to trigger a log entry. If the server implements
       // content-negotiation to support XML clients, I don't want the OpenAPI
       // spec to be converted into an XML response.
-      const settings = OPENAPI_SPEC_MAPPING[request.url];
-      return this._serveOpenApiSpec(request, response, settings);
+      const form = mapping[request.url];
```
Comment (virkt25) [inline packages/rest/src/rest.server.ts]: Why is this called `form`? Seems unintuitive.

File: packages/rest/src/rest.server.ts [hunk 20]
```diff
@@ -150,47 +139,62 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
-      this.sequence(options.sequence);
+    config.openApiSpec = config.openApiSpec || {};
+    config.openApiSpec.endpointMapping =
+      config.openApiSpec.endpointMapping || OPENAPI_SPEC_MAPPING;
+    config.apiExplorer = config.apiExplorer || {};
+
+    config.apiExplorer.url =
+      config.apiExplorer.url || 'https://loopback.io/api-explorer';
+
+    config.apiExplorer.urlForHttp =
```
Comment (virkt25) [inline packages/rest/src/rest.server.ts]: I think `httpUrl` might be a better name.

File: packages/rest/src/rest.server.ts (hunk 16 above)
Comment (virkt25) [reply packages/rest/src/rest.server.ts]: How about `httpUrl`?

File: packages/rest/src/rest.server.ts (hunk 19 above)
Comment (raymondfeng) [reply packages/rest/src/rest.server.ts]: What's a better name for `version + format`?

File: packages/rest/test/integration/rest.server.integration.ts [hunk 21]
```diff
@@ -262,6 +269,119 @@ describe('RestServer (integration)', () => {
...
+      '/openapi.json',
+    );
+    expect(response.body).to.containDeep({
+      openapi: '3.0.0',
+      info: {
+        title: 'LoopBack Application',
+        version: '1.0.0',
+      },
+      servers: [{url: 'http://127.0.0.1:8080'}],
+      paths: {
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: Please simplify this assertion to focus only on the part that's important (i.e. `servers` property) and leave out the rest of the API spec that's only adding noise to error messages and test source code.

```ts
expect(response.body.servers).to.deepEqual([
  {url: 'http://127.0.0.1:8080'},
]);
```

File: packages/rest/test/integration/rest.server.integration.ts [hunk 22]
```diff
@@ -262,6 +269,119 @@ describe('RestServer (integration)', () => {
...
+    const server = await givenAServer({
+      rest: {
+        port: 0,
+        openApiSpec: {
+          servers: [{url: 'http://127.0.0.1:8080'}],
+        },
+      },
+    });
+    const greetSpec = {
+      responses: {
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: ~~This OpenAPI snippet is repeated in many tests. Please extract it into a helper function, e.g a function that adds a dummy route.~~~

Do we really need a route to verify how top-level spec properties are handled? Isn't in enough to render the spec for an empty app? Such app should still produce `servers` property of the spec. See the test case "exposes endpoints with openApiSpec.endpointMapping" for an example.

Leaving out the route we don't need in the test will make the test code much simpler and easier to understand. Please remove it.

File: packages/rest/test/integration/rest.server.integration.ts [hunk 23]
```diff
@@ -359,12 +480,46 @@ servers:
...
+        200: {
+          schema: {type: 'string'},
+          description: 'greeting of the day',
+        },
+      },
+    };
+    server.route(new Route('get', '/greet', greetSpec, function greet() {}));
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: I believe this route is not needed by this test, could you please remove?

File: packages/rest/test/integration/rest.server.integration.ts [hunk 24]
```diff
@@ -262,6 +269,119 @@ describe('RestServer (integration)', () => {
...
+    });
+    const greetSpec = {
+      responses: {
+        200: {
+          content: {'text/plain': {schema: {type: 'string'}}},
+          description: 'greeting of the day',
+        },
+      },
+    };
+    server.route(new Route('get', '/greet', greetSpec, function greet() {}));
```
Comment (bajtos) [inline packages/rest/test/integration/rest.server.integration.ts]: I believe this route is not needed to verify that "setServersFromRequest" is handled correctly, could you please remove it?

**Comment by raymondfeng** [comment]
@bajtos Your comments are addressed:
- Clean up tests
- Add a test for ipv6 url construction
- Cherry-pick the code from #1664 to mount endpoints for OpenAPI specs

PTAL.

File: packages/rest/src/rest.server.ts [hunk 25]
```diff
@@ -151,41 +141,56 @@ export class RestServer extends Context implements Server, HttpServerLike {
...
-      this.sequence(options.sequence);
+    config.openApiSpec = config.openApiSpec || {};
+    config.openApiSpec.endpointMapping =
+      config.openApiSpec.endpointMapping || OPENAPI_SPEC_MAPPING;
+    config.apiExplorer = config.apiExplorer || {};
+
+    config.apiExplorer.url =
+      config.apiExplorer.url || 'https://loopback.io/api-explorer';
+
+    config.apiExplorer.httpUrl =
```
Comment (hacksparrow) [inline packages/rest/src/rest.server.ts]: Can we come up with a more intuitive interface/names for `apiExplorer.url` and `apiExplorer.httpUrl`?

`httpUrl` makes me wonder if there would be `httpsUrl` as well.

I know it is not trivial and might get overly complicated for a simple thing, but I'd like the ability to set a HTTP url and HTTPS url, and mark one as the default.

Comment (hacksparrow) [reply packages/rest/src/rest.server.ts]: On second thoughts, I am fine with this. HTTPS is starting to become the expected default in browsers, let HTTPS be the default `apiExplorer.url`.
